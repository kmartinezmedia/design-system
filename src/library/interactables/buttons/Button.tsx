import React, { FunctionComponent } from 'react';
import {
  GestureResponderEvent,
  Text,
  TouchableWithoutFeedback,
  Animated,
  ActivityIndicator,
  StyleSheet,
  View as RNView,
} from 'react-native';
import { View, ViewProps } from '@designSystem/primitives';
import {
  border,
  fontFamily,
  fontSize,
  fontWeight,
  spacing as spacingConstant,
} from '@designSystem/theme';
import { ButtonVariant } from '@designSystem/types';
import { useButtonVariant, usePressAnimation } from '@designSystem/hooks';

const defaultSpacing = { bottom: 4 } as const;

export interface ButtonProps extends Omit<ViewProps, 'onPress'> {
  testID?: string;
  variant?: ButtonVariant;
  disabled?: boolean;
  isLoading?: boolean;
  outline?: boolean;
  onPress?: (e: GestureResponderEvent) => void;
}

export const Button: FunctionComponent<ButtonProps> = ({
  variant = 'primary',
  spacing = defaultSpacing,
  flexGrow = 0,
  testID,
  children,
  disabled,
  isLoading,
  onPress,
  ...props
}) => {
  const { textColor, ...variantStyles } = useButtonVariant(variant, disabled);
  const [pressIn, pressOut, scale] = usePressAnimation();

  return (
    <View flexGrow={flexGrow} spacing={spacing} {...props}>
      <TouchableWithoutFeedback
        testID={testID}
        onPress={onPress}
        onPressIn={pressIn}
        onPressOut={pressOut}
        disabled={disabled || isLoading}
      >
        <Animated.View
          style={[
            styles.button,
            variant === 'tertiary' && styles.tertiary,
            variantStyles,
            { transform: [{ scale }] },
          ]}
          {...props}
        >
          {isLoading ? (
            <RNView style={styles.loading}>
              <ActivityIndicator size="small" color={textColor} />
            </RNView>
          ) : (
            <Text style={[styles.text, { color: textColor }]}>{children}</Text>
          )}
        </Animated.View>
      </TouchableWithoutFeedback>
    </View>
  );
};

const styles = StyleSheet.create({
  button: {
    alignItems: 'center',
    flexDirection: 'column',
    justifyContent: 'center',
    borderStyle: 'solid',
    width: '100%',
    borderRadius: border.buttonRadius,
    padding: spacingConstant[4],
  },
  tertiary: {
    padding: spacingConstant[1],
  },
  loading: {
    height: 18,
  },
  text: {
    fontFamily: fontFamily.medium,
    fontSize: fontSize.body,
    fontWeight: fontWeight.medium,
  },
});
