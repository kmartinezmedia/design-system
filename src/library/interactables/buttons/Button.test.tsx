import React from 'react';
import 'react-native';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { Button } from './Button';

describe('Button', () => {
  it('default', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <Button>Hey, lil 🐛</Button>
          <Button variant="secondary">Hey, lil 🐛</Button>
          <Button variant="positive">Hey, lil 🐛</Button>
          <Button variant="negative">Hey, lil 🐛</Button>
          <Button variant="tertiary">Hey, lil 🐛</Button>
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
  it('positions', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <Button spacing={3}>Hey, lil 🐛</Button>
          <Button spacing={2} pin="right">
            Hey, lil 🐛
          </Button>
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
});
