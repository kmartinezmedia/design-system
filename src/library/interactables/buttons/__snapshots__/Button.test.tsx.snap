// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "paddingBottom": 16,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1652F0",
            "borderColor": "#1652F0",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 16,
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Graphik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Hey, lil 🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "paddingBottom": 16,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#F1F3F5",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 16,
            "textColor": "#1652F0",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Graphik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#1652F0",
              },
            ]
          }
        >
          Hey, lil 🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "paddingBottom": 16,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#00A45F",
            "borderColor": "#00A45F",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 16,
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Graphik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Hey, lil 🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "paddingBottom": 16,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#E5212E",
            "borderColor": "#00A45F",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 16,
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Graphik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Hey, lil 🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "paddingBottom": 16,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 8,
            "borderStyle": "solid",
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 4,
            "textColor": "#587088",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Graphik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#587088",
              },
            ]
          }
        >
          Hey, lil 🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
]
`;

exports[`Button positions 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "padding": 12,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1652F0",
            "borderColor": "#1652F0",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 16,
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Graphik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Hey, lil 🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "padding": 8,
        },
        Object {
          "bottom": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        pin="right"
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#1652F0",
            "borderColor": "#1652F0",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "padding": 16,
            "textColor": "#FFFFFF",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Graphik-Medium",
                "fontSize": 16,
                "fontWeight": "500",
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Hey, lil 🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
]
`;
